google_compute_engine/distro_lib/ip_forwarding_utils.py

```diff
@@ -0,0 +1,303 @@
+#!/usr/bin/python
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utilities for configuring IP address forwarding."""
+
+import re
+import subprocess
+try:
+  # The following modules are required by IpForwardingUtilsIfconfig.
+  import netifaces
+  import netaddr
+except ImportError:
+  netifaces = None
+  netaddr = None
+
+
+IP_REGEX = re.compile(r'\A(\d{1,3}\.){3}\d{1,3}\Z')
+IP_ALIAS_REGEX = re.compile(r'\A(\d{1,3}\.){3}\d{1,3}/\d{1,2}\Z')
+
+
+class IpForwardingUtils(object):
```

Comment:
So, I wonder if we could make this a bit clearer. A couple thoughts:
- Could we rename "IpForwardingUtils" to "IpForwardingUtilsBase" or something and then have IpForwardingUtils be a dummy class of some sort that will return the correct class based on a constant that we can override or something? (I admit that I'm not 100% sure how to make that work, so if it's not feasible, then no worries.)

A different idea:
In the individual utils.py files, instead of doing the override as part of the import, could we instead do something like this:

from google_compute_engine.distro_utils import ip_forwarding_utils
ip_forwarding_utils.IpForwardingUtils = ip_forwarding_utils.IpForwardingUtilsIfconfig

I believe that the result would be similar to the imports that we are doing now, but it seems like it would be a little clearer about what's going on.

Reply:
What do you think if we instead use a function in the utils.py as the previous commit [1], and use the `__new__` function in the original class `google_compute_engine.networking.ip_forwarding.IpForwardingUtils` which will return an instance created by utils.py ?
This would keep the utils.py with the same pattern (overriding functions).


[1] https://github.com/GoogleCloudPlatform/compute-image-packages/pull/622/commits/1a9dfb323793c11ac7f3b15ec687d665c262e4eb#diff-ffd46ac80f68d2b94cbe7060700b93b5R49

Reply:
Hi. I would also force the `IpForwardingUtils` to be an abstract class (by using the abs module) instead of only empty implementation. That would improve maintainability when extending new classes.

Reply:
To the extent that I understand what you are saying, it seems reasonable. (I haven't used __new__ extensively, so it's hard for me to be 100% confident in that.)

I do agree that setting it up as an abstract class would help in the long run.

Mainly, though, I think the most important thing is just making it so that it's clear what's going on, and the override as part of the import wasn't doing that.